cpp/src/io/parquet/experimental/dictionary_page_filter.cu

```diff
@@ -46,11 +48,1538 @@ using parquet::detail::ColumnChunkDesc;
+ *
+ * @param page_data Pointer to the page data buffer
+ * @param page_data_size Page data buffer size
+ * @param buffer_offset Current offset into the page data buffer
+ * @param error Pointer to the kernel error code
+ * @return Decoded string value
+ */
+__device__ cudf::string_view decode_string_value(uint8_t const* page_data,
+                                                 int32_t page_data_size,
+                                                 int32_t& buffer_offset,
+                                                 kernel_error::pointer error)
+{
+  if (is_stream_overrun(buffer_offset, sizeof(int32_t), page_data_size)) {
+    set_error(error, decode_error::DATA_STREAM_OVERRUN);
+    return {};
+  }
+
+  // Decode string length
+  auto const string_length = static_cast<int32_t>(*(page_data + buffer_offset));
+  buffer_offset += sizeof(int32_t);
+
+  // Check if we have a stream overrun
+  if (is_stream_overrun(buffer_offset, string_length, page_data_size)) {
+    set_error(error, decode_error::DATA_STREAM_OVERRUN);
+    return {};
+  }
+
+  // Decode cudf::string_view value
+  auto const decoded_value =
+    cudf::string_view{reinterpret_cast<char const*>(page_data + buffer_offset),
+                      static_cast<cudf::size_type>(string_length)};
+
+  // Update the buffer offset
+  buffer_offset += string_length;
+
+  return decoded_value;
+}
+
+/**
+ * @brief Decode string column chunk dictionaries and build `cuco::static_set`s from them,
+ * one hash set per dictionary
+ *
+ * @param pages Column chunk dictionary page headers
+ * @param decoded_data Span of storage for decoded values from all dictionaries
+ * @param set_storage Pointer to the start of the bulk cuco hash set slots
+ * @param set_offsets Pointer to offsets into the bulk set storage for each dictionary
+ * @param value_offsets Pointer to offsets into running sum of values in each dictionary
+ * @param num_dictionary_columns Total number of columns with dictionaries
+ * @param dictionary_col_idx Index of the current dictionary column
+ * @param error Pointer to the kernel error code
+ */
+CUDF_KERNEL void __launch_bounds__(DECODE_BLOCK_SIZE)
+  build_string_dictionaries(PageInfo const* const pages,
+                            cudf::device_span<cudf::string_view> decoded_data,
+                            slot_type* const set_storage,
+                            cudf::size_type const* const set_offsets,
+                            cudf::size_type const* const value_offsets,
+                            cudf::size_type total_row_groups,
+                            cudf::size_type num_dictionary_columns,
+                            cudf::size_type dictionary_col_idx,
+                            kernel_error::pointer error)
+{
+  // Each thread decodes one dictionary page and inserts into the corresponding cuco hash set
+
+  // Current row group index
+  for (size_type row_group_idx = cudf::detail::grid_1d::global_thread_id();
```

Comment:
Using one thread per dictionary page instead of (thread0 of) each warp per dictionary